src/cluster_legacy.c

```diff
@@ -3986,28 +4163,39 @@ int clusterSendModuleMessageToTarget(const char *target,
  * Otherwise:
  * Publish this message across the slot (primary/replica).
  * -------------------------------------------------------------------------- */
-void clusterPropagatePublish(robj *channel, robj *message, int sharded) {
-    clusterMsgSendBlock *msgblock;
+void clusterPropagatePublish(robj *channel, robj **messages, int count, int sharded) {
+    /* Currently publish command does not support multiple payload. */
+    serverAssert(count == 1);
 
-    if (!sharded) {
-        msgblock = clusterCreatePublishMsgBlock(channel, message, CLUSTERMSG_TYPE_PUBLISH);
-        clusterBroadcastMessage(msgblock);
-        clusterMsgSendBlockDecrRefCount(msgblock);
-        return;
+    clusterMsgSendBlock *msgblock, *msgblock_light;
+
+    msgblock_light = clusterCreatePublishLightMsgBlock(
+        channel, messages, count, sharded ? CLUSTERMSG_TYPE_PUBLISHSHARD_LIGHT : CLUSTERMSG_TYPE_PUBLISH_LIGHT);
+
+    ClusterNodeIterator iter;
+    if (sharded) {
+        list *nodes_for_slot = clusterGetNodesInMyShard(server.cluster->myself);
+        serverAssert(nodes_for_slot != NULL);
+        clusterNodeIterInitMyShard(&iter, nodes_for_slot);
+    } else {
+        clusterNodeIterInitAllNodes(&iter, server.cluster->nodes);
     }
 
-    listIter li;
-    listNode *ln;
-    list *nodes_for_slot = clusterGetNodesInMyShard(server.cluster->myself);
-    serverAssert(nodes_for_slot != NULL);
-    listRewind(nodes_for_slot, &li);
-    msgblock = clusterCreatePublishMsgBlock(channel, message, CLUSTERMSG_TYPE_PUBLISHSHARD);
-    while ((ln = listNext(&li))) {
-        clusterNode *node = listNodeValue(ln);
+    clusterNode *node;
+    while ((node = clusterNodeIterNext(&iter)) != NULL) {
         if (node->flags & (CLUSTER_NODE_MYSELF | CLUSTER_NODE_HANDSHAKE)) continue;
-        clusterSendMessage(node->link, msgblock);
+        /* If the cluster is not marked as homogeneous send respective message blocks to all nodes. */
+        if (nodeSupportsLightMsgHdr(node)) {
+            clusterSendMessage(node->link, msgblock_light);
+        } else {
+            msgblock = clusterCreatePublishMsgBlock(channel, messages[0],
+                                                    sharded ? CLUSTERMSG_TYPE_PUBLISHSHARD : CLUSTERMSG_TYPE_PUBLISH);
+            clusterSendMessage(node->link, msgblock);
```

Comment:
Here we create a new msgblock for every node that needs it. Do you think it will be very rare that more than one node needs it?

I was thinking we initialize it to NULL first, and we create it here only if msgblock == NULL, so we only create it at most once. That's what I meant by lazy initialization. After the loop, we can call clusterMsgSendBlockDecrRefCount only if msgblock != NULL.

Do you think this would be too complex?

Reply:
I did think of it, but this heterogenous cluster will be a temporal thing. Once the cluster supports light header we are good.

Also how often would an old node be added to the cluster?

Reply:
Okay but lets add it

Reply:
I guess it's only during rolling upgrade. But I believe we support it, so it's not forbidden. I don't know if why anyone would do it though, except during upgrade.

Reply:
YEs thats what I was also thinking